src/main/java/com/google/devtools/build/lib/rules/proto/ProtoCommon.java

```diff
@@ -404,44 +399,57 @@ private static Library createLibraryWithVirtualSourceRootMaybe(
+    NestedSetBuilder<Pair<PathFragment, ImmutableList<Artifact>>> protos =
+        NestedSetBuilder.stableOrder();
+    protos.add(new Pair<>(PathFragment.create(library.getSourceRoot()), library.getSources()));
+    for (ProtoInfo info : ruleContext.getPrerequisites("deps", TARGET, ProtoInfo.PROVIDER)) {
+      protos.addTransitive(info.getExportedProtos());
     }
-
-    return protoSourceRoots.build();
+    return protos.build();
   }
 
   /**
-   * Returns a set of the {@code proto_source_root} collected from the current library and the
-   * direct dependencies.
-   *
-   * <p>Assumes {@code currentProtoSourceRoot} is the same as the package name.
+   * Returns a set of pairs of {@code proto_source_root -> [sources]} of `.proto` files that are
+   * exported by a {@code proto_library} target.
    */
-  private static NestedSet<String> computeStrictImportableProtoSourceRoots(
-      RuleContext ruleContext, String currentProtoSourceRoot) {
-    return getProtoSourceRootsOfAttribute(ruleContext, currentProtoSourceRoot, "deps");
+  private static NestedSet<Pair<PathFragment, ImmutableList<Artifact>>> computeExportedProtos(
+      RuleContext ruleContext, Library library) {
+    NestedSetBuilder<Pair<PathFragment, ImmutableList<Artifact>>> protos =
+        NestedSetBuilder.stableOrder();
+    protos.add(new Pair<>(PathFragment.create(library.getSourceRoot()), library.getSources()));
+    for (ProtoInfo info : ruleContext.getPrerequisites("exports", TARGET, ProtoInfo.PROVIDER)) {
+      protos.addTransitive(info.getExportedProtos());
+    }
+    if (library.getSources().isEmpty()) {
+      // This is an alias library (i.e. a `proto_library` target without `srcs`). Treat all exports
+      // as deps.
+      // TODO(yannic): Consider requiring users to use `exports` in this case.
+      for (ProtoInfo info : ruleContext.getPrerequisites("deps", TARGET, ProtoInfo.PROVIDER)) {
+        protos.addTransitive(info.getExportedProtos());
+      }
+    }
+    return protos.build();
   }
 
   /**
-   * Returns a set of the {@code proto_source_root} collected from the current library and the
-   * exported dependencies.
-   *
-   * <p>Assumes {@code currentProtoSourceRoot} is the same as the package name.
+   * Returns a set of pairs of {@code proto_source_root -> [sources]} of `.proto` files that are
+   * reachable by a {@code proto_library} target.
    */
-  private static NestedSet<String> computeExportedProtoSourceRoots(
-      RuleContext ruleContext, String currentProtoSourceRoot) {
-    return getProtoSourceRootsOfAttribute(ruleContext, currentProtoSourceRoot, "exports");
+  private static NestedSet<Pair<PathFragment, ImmutableList<Artifact>>> computeTransitiveProtos(
+      RuleContext ruleContext,
+      NestedSet<Pair<PathFragment, ImmutableList<Artifact>>> exportedProtos) {
+    NestedSetBuilder<Pair<PathFragment, ImmutableList<Artifact>>> protos =
+        NestedSetBuilder.stableOrder();
+    protos.addTransitive(exportedProtos);
```

Comment:
This is inconsistent with how `transitiveProtoSources` (the `NestedSet<Artifact>`) is computed: that is only local sources + deps and this one is local sources + deps + exports. Make this consistent (or else explain in a comment why the discrepancy exists)